### PR TITLE
WIP: Implement smoothing operations

### PR DESCRIPTION
**Pull request by drdavella**
This PR represents the work that @robelgeda and @brechmos-stsci have done with respect to smoothing operations. This builds the work done in #55.

### CODE REVIEW

File: cubeviz/tools/smoothing.py [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+    output_component_id="Cube", output_as_component=False,
+    original_data=None):
+    if output_as_component:
+        original_data.add_component(cube._data, output_component_id)
+        original_data.broadcast(
+            msg.DataAddComponentMessage(
+                original_data,
+                output_component_id
+            )
+        )
```
Comment (astrofrog) [inline cubeviz/tools/smoothing.py]: I think the DataAddComponentMesssage should be automatically emitted when calling ``add_component`` above - if not, that's a bug in glue.

File: cubeviz/tools/smoothing.py [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+        new_component = Component(cube._data)
```
Comment (astrofrog) [inline cubeviz/tools/smoothing.py]: What is the purpose of ``new_component``?

File: cubeviz/layout.py [hunk 3]
```diff
@@ -242,7 +250,17 @@ def _dict_to_menu(self, menu_dict):
         return menu_widget
 
     def _open_dialog(self, name, widget):
-        get_text(name, "What's your name?")
+        #get_text(name, "What's your name?")
```
Comment (brechmos-stsci) [inline cubeviz/layout.py]: @robelgeda Could you remove this commented out line of text?  Not needed now.

File: cubeviz/layout.py [hunk 4]
```diff
@@ -242,7 +250,17 @@ def _dict_to_menu(self, menu_dict):
...
+        if name == "Filter":
+            ex = smoothing_gui.SelectSmoothing(
+                self._data, 
+                self.session.data_collection,
+                parent=self#self.session.application
```
Comment (brechmos-stsci) [inline cubeviz/layout.py]: same thing, probably best to remove the comment

File: cubeviz/tools/smoothing.py [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+        "MexicanHat": {"spatial" : convolution.MexicanHat2DKernel, "spectral" : convolution.MexicanHat1DKernel},
+        "Trapezoid": {"spectral" : convolution.Trapezoid1DKernel},
+        "TrapezoidDisk": {"spatial" : convolution.TrapezoidDisk2DKernel},
+        "AiryDisk": {"spatial" : convolution.AiryDisk2DKernel},
+        "TopHat" :{"spatial" : convolution.Tophat2DKernel},
+        "Median" : {"spatial" : None, "spectral" : None}
+    }
+    return kernel_registry
+
+def print_kernel_types():
```
Comment (brechmos-stsci) [inline cubeviz/tools/smoothing.py]: If this function was just used for debugging we can probably delete it.

**Comment by drdavella** [comment]
Does anyone mind if this gets rebased or should we wait until the review is finished?

File: cubeviz/tools/smoothing.py [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+        If missing smoothing functions in spectral_cube.SpectralCube.
+        User is informed to update their spectral_cube.
+    TypeError: 
+        If input kernel is not numpy.ndarray nor from 
+        astropy.convolution.
+    Exception: 
+        If input numpy.ndarray kernel's numpy.ndarray.size 
+        is not supported.
+    """
+    print("In")
```
Comment (drdavella) [inline cubeviz/tools/smoothing.py]: Looks like a leftover from development

File: cubeviz/tools/smoothing.py [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+    if kernel is None:
+        if smoothing_axis is None:
+            smoothing_axis = "spatial"
+        else:
+            smoothing_axis = smoothing_axis
+
+        kernel_registry = get_kernel_registry()
+
+        kernel_type_list = [i for i in kernel_registry.keys()]
+        kernel_type = kernel_type
```
Comment (drdavella) [inline cubeviz/tools/smoothing.py]: Should be deleted

**Comment by brechmos-stsci** [comment]
@robelgeda @javerbukh and I went through the PR together. There are a few things @robelgeda is going to change around and (prob tomorrow or so).

**Comment by astrofrog** [comment]
As promised, here is a proof of concept of doing smoothing on-the-fly: https://github.com/spacetelescope/cubeviz/pull/122

**Comment by robelgeda** [comment]
Replaced by #128

**Pull request closed without merging**